config/src/main/java/com/typesafe/config/impl/ConfigNodeKeyValue.java

```diff
@@ -0,0 +1,48 @@
+package com.typesafe.config.impl;
+
+import com.typesafe.config.ConfigNode;
+
+import java.util.ArrayList;
+import java.util.Collection;
+
+public class ConfigNodeKeyValue implements ConfigNode{
+    final private ArrayList<ConfigNode> children;
+    private int configNodeValueIndex;
+    private ConfigNodeKey key;
+    private ConfigNodeValue value;
```

Comment:
since these three fields are all caches (they could be computed on the fly), I'd default to computing them on the fly, and only cache if we have to.